compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ForLoopsLowering.kt

```diff
@@ -117,24 +117,53 @@ private class RangeLoopTransformer(
 ) : IrElementTransformerVoidWithContext() {
 
     private val symbols = context.ir.symbols
-    private val iteratorToLoopHeader = mutableMapOf<IrVariableSymbol, ForLoopHeader>()
-    private val headerInfoBuilder = HeaderInfoBuilder(context, this::getScopeOwnerSymbol)
+    private val headerInfoBuilder = DefaultHeaderInfoBuilder(context, this::getScopeOwnerSymbol)
     private val headerProcessor = HeaderProcessor(context, headerInfoBuilder, this::getScopeOwnerSymbol)
 
     fun getScopeOwnerSymbol() = currentScope!!.scope.scopeOwnerSymbol
 
-    override fun visitVariable(declaration: IrVariable): IrStatement {
-        val initializer = declaration.initializer
-        if (initializer == null || initializer !is IrCall) {
-            return super.visitVariable(declaration)
+    override fun visitBlock(expression: IrBlock): IrExpression {
+        // LoopExpressionGenerator in psi2ir lowers `for (loopVar in <someIterable>) { // Loop body }` into an IrBlock with origin FOR_LOOP.
+        // This block has 2 statements:
+        //
+        //   // #1: The "header"
+        //   val it = <someIterable>.iterator()
+        //
+        //   // #2: The inner while loop
+        //   while (it.hasNext()) {
+        //     val loopVar = it.next()
+        //     // Loop body
+        //   }
+        //
+        // We primarily need to determine HOW to optimize the for loop from the iterable expression in the header (e.g., if it's a
+        // `withIndex()` call, a progression such as `10 downTo 1`). However in some cases (e.g., for `withIndex()`), we also need to
+        // examine the while loop to determine if we CAN optimize the loop.
+        if (expression.origin != IrStatementOrigin.FOR_LOOP) {
+            return super.visitBlock(expression)  // Not a for-loop block.
+        }
+
+        with(expression.statements) {
+            assert(size == 2)
```

Comment:
Please add assertion message

Reply:
...and maybe element `dump `

Reply:
@max-kammerer Done. Also added for other asserts.